Classes/MonomeArc.sc

```diff
@@ -0,0 +1,343 @@
+						OSCdef(("deltaFunc_" ++ dvcID).asSymbol).free;
+						oscout = NetAddr.new("localhost", portlst[devicenum].value);
+						Post << "MonomeArc: using device on port #" << portlst[devicenum].value << Char.nl;
+
+						for(0, 3, { arg i; oscout.sendMsg(prefixID++"/ring/all", i, 0);});
+
+						prefixes[devicenum] = prefixID;
+
+						prefixDiscover.free;
+						prefixDiscover = OSCdef.newMatching(\monomeprefix,
+							{|msg, time, addr, recvPort|
+								prefixes[devicenum] = prefixID;
+						}, '/sys/prefix', oscout);
+
+						oscout.sendMsg("/sys/port", NetAddr.localAddr.port);
+						oscout.sendMsg("/sys/prefix", prefixID);
+						oscout.sendMsg("/sys/rotation", rot);
+						oscout.sendMsg("/sys/info");
+
+						// collect individual LED messages into a 'map':
+						quadDirty[dvcID] = Array.fill(4,{0});
+						ledQuads[dvcID] = Array.fill(4,{Array.fill(64,{0})});
+
+						addCallbackComplete[dvcID] = false;
+						addCallback.value(registeredDevices[dvcID], portlst[dvcID], prefixes[dvcID]);
+						seroscnet.sendMsg("/serialosc/notify", "127.0.0.1", NetAddr.localAddr.port);
+					},
+					{
+						("!! no monome arc detected at device slot " ++ devicenum).warn;
+					}
+				);
+			},
+			{
+				("no monome arc detected at device slot " ++ devicenum).warn;
+			}
+		);
+	}
+
+	usePort { arg portnum;
+		dvcID = portlst.indexOf(portnum);
+		oscout = NetAddr.new("localhost", portnum);
+		Post << "MonomeArc: using device # " << dvcID << Char.nl;
+
+		oscout.sendMsg("/sys/port", NetAddr.localAddr.port);
+		oscout.sendMsg("/sys/prefix", prefixID);
+		oscout.sendMsg("/sys/rotation", rot);
+	}
+
+	delta { arg func;
+		deltaFunc = OSCdef.newMatching(
+			("deltaFunc_" ++ dvcID).asSymbol,
+			{ arg message, time, addr, recvPort;
+				var n = message[1], d = message[2];
+				tick[n] = tick[n] + d;
+				if( dvcID.notNil,{
+					if( this.port.value() == addr.port, {
+						if( tick[n].abs >= sens[n], {
+							var val = tick[n]/sens[n];
+							if( val > 0,
+								{val = val.floor},
+								{val = val.ceil}
+							);
+							func.value(n, val);
+							tick[n] = tick[n].mod(sens[n]);
+						});
+					});
+				});
+			},
+			prefixID++"/enc/delta"
+		);
+	}
+
+	key { arg func;
+		keyFunc = OSCdef.newMatching(
+			("keyFunc_" ++ dvcID).asSymbol,
+			{ arg message, time, addr, recvPort;
+				var n = message[1], s = message[2];
+				if( dvcID.notNil,{
+					if( this.port.value() == addr.port, {
+						func.value(n, s);
+					});
+				});
+			},
+			prefixID++"/enc/key"
+		);
+	}
+
+	refresh {
+		for (0, 3, {
+			arg i;
+			if(quadDirty[dvcID][i] != 0,
+				{
+					oscout.sendMsg(
+						prefixID++"/ring/map",
+						i,
+						*ledQuads[dvcID][i]
+					);
+					quadDirty[dvcID][i] = 0;
+				}
+			);
+		});
+
+	}
+
+	led { | ring, led, val |
+		ledQuads[dvcID][ring][(led + scaleFactor).wrap(0, 63)] = val;
+		quadDirty[dvcID][ring] = 1;
+	}
+
+	all { | ring, val |
+		for( 0, 63, {
+			arg i;
+			this.led(ring, i, val);
+		});
+	}
+
+	dark {
```

Comment:
i'd rename this `alloff` to be perfectly clear (sorry i know it's more boring)

Reply:
leftover, will clean!